### PR TITLE
Add support for user-defined Fairness measures and user-defined Losses

### PR DESCRIPTION
**Pull request by mperrot**
Summary of the backward compatible changes:

- Added support for user-defined Losses by refactoring the way losses are handled. A base class called FairnessLoss that takes a torch loss as an argument has been created. Added a wrapper for CrossEntropyLoss for backward compatibility.
- Added support for user-defined Fairness measures that inherit the FairnessMeasure base class.
- Added basic unittesting for the different fairness functions.

Summary of the non-backward compatible changes:
- Removed the support for the mode parameter in the forward function.

Bug fix:
- Fixed a bug in Accuracy Parity (in the method init_C) when the number of classes was higher than 2.

### CODE REVIEW

**Comment by saist1993** [comment]
Hey! This is really good. I like the abstraction. However, before merging I have following comments:

- **Mode Parameter Support**: We need some form of mechanism to signal the cross entropy loss not to update fairness weights while training. This is because we use the same loss function for train and test split. If there does not exists such mechanism, we will also update the fairness weights while testing.  One could also accomplish the same by defining two loss function. However, I think this increases the cognitive load and has a potential for hard to figure out bugs. Let me know, if I misunderstood something here. 

- **Split cross entropy file**: I would suggest to split the ```cross_entropy.py``` file into two files. One which exclusively defines fairness loss and another which defines cross entropy loss.

**Comment by mperrot** [comment]
Okay, I added the Mode Parameter support back as it would indeed be a problem to update the weights in validation or test steps. To keep it backward compatible, the default is "train". Note that one has to explicitly change the mode to something else to compute the weighted loss without updating the weights.
